server/datastore/mysql/host_certificates.go

```diff
@@ -0,0 +1,197 @@
+			return ctxerr.Wrap(ctx, err, "insert host certs")
+		}
+		if err := softDeleteHostCertsDB(ctx, tx, hostID, toDelete); err != nil {
+			return ctxerr.Wrap(ctx, err, "soft delete host certs")
+		}
+		return nil
+	})
+}
+
+func listHostCertsDB(ctx context.Context, tx sqlx.QueryerContext, hostID uint, opts fleet.ListOptions) ([]*fleet.HostCertificateRecord, *fleet.PaginationMetadata, error) {
+	// TODO: move this to the service layer and do validation of the order key?
+	if opts.OrderKey == "" {
+		// default sort by common name ascending
+		opts.OrderKey = "common_name"
+		opts.OrderDirection = fleet.OrderAscending
+	}
+
+	stmt := `
+SELECT 
+	id,
+	sha1_sum,
+	host_id,
+	created_at,
+	deleted_at,
+	not_valid_before,
+	not_valid_after,
+	certificate_authority,
+	common_name,
+	key_algorithm,
+	key_strength,
+	key_usage,
+	serial,
+	signing_algorithm,
+	subject_country,
+	subject_org,
+	subject_org_unit,
+	subject_common_name,
+	issuer_country,
+	issuer_org,
+	issuer_org_unit,
+	issuer_common_name
+FROM
+	host_certificates 
+WHERE 
+	host_id = ?
+	AND deleted_at IS NULL`
+
+	args := []interface{}{hostID}
+	stmtPaged, args := appendListOptionsWithCursorToSQL(stmt, args, &opts)
+
+	var certs []*fleet.HostCertificateRecord
+	if err := sqlx.SelectContext(ctx, tx, &certs, stmtPaged, args...); err != nil {
+		return nil, nil, err
+	}
+
+	var metaData *fleet.PaginationMetadata
+	if opts.IncludeMetadata {
+		metaData = &fleet.PaginationMetadata{HasPreviousResults: opts.Page > 0}
+		if len(certs) > int(opts.PerPage) { //nolint:gosec // dismiss G115
+			metaData.HasNextResults = true
+			certs = certs[:len(certs)-1]
+		}
+	}
+
+	return certs, metaData, nil
+}
+
+func insertHostCertsDB(ctx context.Context, tx sqlx.ExtContext, certs []*fleet.HostCertificateRecord) error {
+	if len(certs) == 0 {
+		return nil
+	}
+
+	stmt := `
+INSERT INTO host_certificates (
+	host_id,
+	sha1_sum,
+	not_valid_before,
+	not_valid_after,
+	certificate_authority,
+	common_name,
+	key_algorithm,
+	key_strength,
+	key_usage,
+	serial,
+	signing_algorithm,
+	subject_country,
+	subject_org,
+	subject_org_unit,
+	subject_common_name,
+	issuer_country,
+	issuer_org,
+	issuer_org_unit,
+	issuer_common_name
+) VALUES %s`
+
+	placeholders := make([]string, 0, len(certs))
+	args := make([]interface{}, 0, len(certs)*19)
+	for _, cert := range certs {
+		placeholders = append(placeholders, "(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
+		args = append(args,
+			cert.HostID, cert.SHA1Sum, cert.NotValidBefore, cert.NotValidAfter, cert.CertificateAuthority, cert.CommonName,
+			cert.KeyAlgorithm, cert.KeyStrength, cert.KeyUsage, cert.Serial, cert.SigningAlgorithm,
+			cert.SubjectCountry, cert.SubjectOrganization, cert.SubjectOrganizationalUnit, cert.SubjectCommonName,
+			cert.IssuerCountry, cert.IssuerOrganization, cert.IssuerOrganizationalUnit, cert.IssuerCommonName)
+	}
+
+	stmt = fmt.Sprintf(stmt, strings.Trim(strings.Join(placeholders, ","), ","))
```

Comment:
Just a FYI as there's no harm as-is: `strings.Join` places the separator only _between_ elements, not at the end, so you don't need the trim when using Join.